In [1]:
# 1-Charger le dataset avec Python

# Importer les bibliothèques nécessaires
import pandas as pd
import numpy as np
from sklearn.preprocessing import LabelEncoder, StandardScaler
from datetime import datetime
import os

from google.colab import files
uploaded = files.upload()

df = pd.read_csv("data_credit_risk.csv")

# Afficher les premières lignes pour vérifier
print(df.head())

Saving data_credit_risk.csv to data_credit_risk (4).csv
   customer_id       full_name                              email  \
0  CUST-000001     Boubacar Ba     boubacar.ba1@example-bank.test   
1  CUST-000002    Mariam Ahmed    mariam.ahmed2@example-bank.test   
2  CUST-000003  Khadijetou Sow  khadijetou.sow3@example-bank.test   
3  CUST-000004   Ahmed Mokhtar   ahmed.mokhtar4@example-bank.test   
4  CUST-000005   Fatimata Mint   fatimata.mint5@example-bank.test   

            phone  national_id  age gender      region marital_status  \
0  +222-47-892243   9153034861   44      M  Nouakchott         single   
1  +222-35-128419   8418245412   35      M       Adrar        married   
2  +222-39-528873   4186896021   28      F  Nouakchott        married   
3  +222-39-501484   1437200349   47      F  Nouadhibou       divorced   
4  +222-34-223152   9194744247   45      M      Gorgol        married   

  education_level  ... loan_amount loan_duration_months  existing_credits  \
0        bach

**2-Comprendre le dataset (Dataset Description)**

In [2]:
df.info()
df.describe()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3200 entries, 0 to 3199
Data columns (total 25 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   customer_id           3200 non-null   object 
 1   full_name             3200 non-null   object 
 2   email                 3200 non-null   object 
 3   phone                 3200 non-null   object 
 4   national_id           3200 non-null   int64  
 5   age                   3200 non-null   int64  
 6   gender                3200 non-null   object 
 7   region                3200 non-null   object 
 8   marital_status        3200 non-null   object 
 9   education_level       3200 non-null   object 
 10  employment_status     3200 non-null   object 
 11  housing_status        3200 non-null   object 
 12  annual_income         3200 non-null   float64
 13  savings_balance       3200 non-null   float64
 14  checking_balance      3200 non-null   float64
 15  loan_amount          

,national_id,age,annual_income,savings_balance,checking_balance,loan_amount,loan_duration_months,existing_credits,past_due_count,credit_history_score,debt_ratio
count,3.200000e+03,3200.000000,3200.000000,3200.000000,3200.000000,3200.000000,3200.000000,3200.000000,3200.000000,3200.000000,3200.000000
mean,5.510391e+09,38.459063,32224.469228,6294.975384,2640.641203,12058.081203,26.683125,0.963125,0.460313,612.488438,0.645585
std,2.598197e+09,10.655958,20619.081204,7271.934690,3570.235151,7216.875924,14.965947,0.986253,0.673938,78.813187,0.431937
min,1.001547e+09,18.000000,4000.000000,0.000000,-1500.000000,1298.680000,6.000000,0.000000,0.000000,300.000000,0.010000
25%,3.253607e+09,31.000000,17709.590000,1148.572500,345.362500,7109.000000,12.000000,0.000000,0.000000,562.000000,0.340000
50%,5.570581e+09,38.000000,27587.080000,4212.440000,1769.520000,10395.050000,24.000000,1.000000,0.000000,612.000000,0.515000
75%,7.745921e+09,46.000000,41713.907500,8687.730000,3973.167500,14952.952500,36.000000,2.000000,1.000000,667.250000,0.807000
max,9.997613e+09,75.000000,120000.000000,63602.300000,30789.630000,81907.220000,60.000000,6.000000,4.000000,850.000000,1.800000


## CONCLUSION DE L'ANALYSE DU DATASET

### 1. Structure générale du dataset

- Le dataset contient **3 200 lignes** (enregistrements) et **25 colonnes** (variables)
- Aucune valeur manquante n'est présente (toutes les colonnes ont 3200 valeurs non-nulles)

---

### 2. Répartition des colonnes selon leurs types de données

| Type de données | Nombre de colonnes | Colonnes concernées |
|-----------------|-------------------|---------------------|
| **Numerical (int64)** | 6 | `national_id`, `age`, `loan_duration_months`, `existing_credits`, `past_due_count`, `credit_history_score` |
| **Numerical (float64)** | 5 | `annual_income`, `savings_balance`, `checking_balance`, `loan_amount`, `debt_ratio` |
| **Categorical/Object** | 13 | `customer_id`, `full_name`, `email`, `phone`, `gender`, `region`, `marital_status`, `education_level`, `employment_status`, `housing_status`, `loan_purpose`, `application_date` |
| **Boolean (bool)** | 1 | `consent_flag` |
| **Total** | **25** | |

---

### 3. Analyse des variables numériques (via df.describe())

| Statistique | Âge | Revenu annuel | Montant du prêt | Durée du prêt | Dette |
|-------------|-----|---------------|-----------------|---------------|-------|
| **Minimum** | 18 ans | 4 000 | 1 298 | 6 mois | 0.01 |
| **Maximum** | 75 ans | 120 000 | 81 907 | 60 mois | 1.80 |
| **Moyenne** | 38.5 ans | 32 224 | 12 058 | 26.7 mois | 0.65 |
| **Médiane** | 38 ans | 27 587 | 10 395 | 24 mois | 0.52 |

**Observations clés :**
- Les clients ont entre **18 et 75 ans**, avec un âge moyen de **38.5 ans**
- Le revenu annuel varie de **4 000 à 120 000**, moyenne de **32 224**
- Le montant du prêt varie de **1 298 à 81 907**, moyenne de **12 058**
- La durée du prêt s'étend de **6 à 60 mois**, moyenne de **26.7 mois**
- Le ratio d'endettement (`debt_ratio`) varie de **0.01 à 1.80**, moyenne de **0.65**

---

### 4. Variable cible

- La variable cible est **`credit_risk`** (type `object`)
- Elle représente le **risque de crédit** des clients (probablement "Good" ou "Bad")
- Il s'agit donc d'un problème de **classification binaire**


**3-Identifier les données sensibles**

### A) IDENTIFIANTS DIRECTS

| Colonne | Description |
|---------|-------------|
| `customer_id` | Identifiant unique du client |
| `full_name` | Nom complet du client |
| `email` | Adresse email du client |
| `phone` | Numéro de téléphone du client |
| `national_id` | Numéro d'identité nationale du client |

**5 identifiants directs présents**

---

### B) QUASI-IDENTIFIANTS

| Colonne | Description |
|---------|-------------|
| `age` | Âge du client |
| `gender` | Genre du client |
| `region` | Région géographique du client |
| `marital_status` | Situation matrimoniale du client |
| `education_level` | Niveau d'éducation du client |
| `employment_status` | Situation professionnelle du client |
| `housing_status` | Statut de logement du client |

**7 quasi-identifiants présents**

---

### C) DONNÉES FINANCIÈRES SENSIBLES

| Colonne | Description |
|---------|-------------|
| `annual_income` | Revenu annuel du client |
| `savings_balance` | Solde du compte d'épargne |
| `checking_balance` | Solde du compte courant |
| `loan_amount` | Montant du prêt demandé |
| `loan_duration_months` | Durée du prêt en mois |
| `existing_credits` | Nombre de crédits existants |
| `debt_ratio` | Ratio d'endettement |

**7 données financières sensibles présentes**

---

### RÉCAPITULATIF

| Catégorie | Colonnes concernées |
|-----------|---------------------|
| **Identifiants directs** | `customer_id`, `full_name`, `email`, `phone`, `national_id` |
| **Quasi-identifiants** | `age`, `gender`, `region`, `marital_status`, `education_level`, `employment_status`, `housing_status` |
| **Données financières** | `annual_income`, `savings_balance`, `checking_balance`, `loan_amount`, `loan_duration_months`, `existing_credits`, `debt_ratio` |

**4-NETTOYAGE DES DONNÉES :**

**4-1-Supprimer les valeurs incorrectes**

In [3]:
print("=== ÉTAPE 4-1 : Suppression des valeurs incorrectes ===\n")

# 1. Vérifier les valeurs aberrantes dans les colonnes numériques
colonnes_numeriques = df.select_dtypes(include=['int64', 'float64']).columns

print("Valeurs aberrantes détectées :")
for col in colonnes_numeriques:
    Q1 = df[col].quantile(0.25)
    Q3 = df[col].quantile(0.75)
    IQR = Q3 - Q1
    lower_bound = Q1 - 3 * IQR
    upper_bound = Q3 + 3 * IQR

    outliers = df[(df[col] < lower_bound) | (df[col] > upper_bound)]
    if len(outliers) > 0:
        print(f"  - {col} : {len(outliers)} valeurs aberrantes")
        # Remplacer les outliers par la médiane
        mediane = df[col].median()
        df.loc[df[col] < lower_bound, col] = mediane
        df.loc[df[col] > upper_bound, col] = mediane
        print(f"    Remplacées par la médiane ({mediane})")

# 2. Vérifier les valeurs incorrectes dans les colonnes catégorielles
colonnes_categoriques = df.select_dtypes(include=['object']).columns

print("\nValeurs incorrectes dans les colonnes catégorielles :")
for col in colonnes_categoriques:
    valeurs_uniques = df[col].unique()
    print(f"  - {col} : {len(valeurs_uniques)} valeurs uniques")
    # Afficher les valeurs uniques pour vérification
    if len(valeurs_uniques) <= 20:
        print(f"    {valeurs_uniques.tolist()}")

print(f"\n Dataset après suppression des valeurs incorrectes : {df.shape[0]} lignes, {df.shape[1]} colonnes")

=== ÉTAPE 4-1 : Suppression des valeurs incorrectes ===

Valeurs aberrantes détectées :
  - annual_income : 24 valeurs aberrantes
    Remplacées par la médiane (27587.08)
  - savings_balance : 39 valeurs aberrantes
    Remplacées par la médiane (4212.4400000000005)
  - checking_balance : 33 valeurs aberrantes
    Remplacées par la médiane (1769.52)
  - loan_amount : 31 valeurs aberrantes
    Remplacées par la médiane (10395.05)

Valeurs incorrectes dans les colonnes catégorielles :
  - customer_id : 3200 valeurs uniques
  - full_name : 196 valeurs uniques
  - email : 3200 valeurs uniques
  - phone : 3200 valeurs uniques
  - gender : 2 valeurs uniques
    ['M', 'F']
  - region : 12 valeurs uniques
    ['Nouakchott', 'Adrar', 'Nouadhibou', 'Gorgol', 'Inchiri', 'Guidimakha', 'Trarza', 'Brakna', 'Hodh El Gharbi', 'Assaba', 'Tagant', 'Hodh Ech Chargui']
  - marital_status : 4 valeurs uniques
    ['single', 'married', 'divorced', 'widowed']
  - education_level : 5 valeurs uniques
    ['bache

**4-2-Convertir les types de données :**

In [4]:
print("\n=== ÉTAPE 4-2 : Conversion des types de données ===\n")

print("Types avant conversion :")
print(df.dtypes)

# Convertir la date
if 'application_date' in df.columns:
    df['application_date'] = pd.to_datetime(df['application_date'])
    print("\n application_date convertie en datetime")

# Convertir les booléens
if 'consent_flag' in df.columns:
    df['consent_flag'] = df['consent_flag'].astype(bool)
    print(" consent_flag convertie en bool")

# Convertir les colonnes catégorielles en 'category'
colonnes_categoriques = df.select_dtypes(include=['object']).columns
for col in colonnes_categoriques:
    if col != 'credit_risk':
        df[col] = df[col].astype('category')
        print(f" {col} convertie en category")

print("\nTypes après conversion :")
print(df.dtypes)


=== ÉTAPE 4-2 : Conversion des types de données ===

Types avant conversion :
customer_id              object
full_name                object
email                    object
phone                    object
national_id               int64
age                       int64
gender                   object
region                   object
marital_status           object
education_level          object
employment_status        object
housing_status           object
annual_income           float64
savings_balance         float64
checking_balance        float64
loan_amount             float64
loan_duration_months      int64
existing_credits          int64
past_due_count            int64
credit_history_score      int64
debt_ratio              float64
loan_purpose             object
consent_flag               bool
application_date         object
credit_risk              object
dtype: object

 application_date convertie en datetime
 consent_flag convertie en bool
 customer_id convertie en category

**4-3 Encoder les variables catégorielles :**

In [5]:
print("\n=== ÉTAPE 4-3 : Encodage des variables catégorielles ===\n")

# Créer une copie du dataset pour l'encodage
df_encoded = df.copy()

# 1. Encoder la variable cible 'credit_risk'
print("1. Encodage de la variable cible 'credit_risk' :")
le_target = LabelEncoder()
df_encoded['credit_risk_encoded'] = le_target.fit_transform(df_encoded['credit_risk'])
print(f"   - good → {le_target.transform(['good'])[0]}")
print(f"   - bad → {le_target.transform(['bad'])[0]}")
print(f"    credit_risk encodée")

# 2. Encoder les variables catégorielles binaires
print("\n2. Encodage des variables binaires :")
colonnes_binaires = ['gender', 'marital_status', 'housing_status', 'consent_flag']
for col in colonnes_binaires:
    if col in df_encoded.columns:
        le = LabelEncoder()
        df_encoded[col + '_encoded'] = le.fit_transform(df_encoded[col])
        print(f"    {col} encodée (valeurs: {dict(zip(le.classes_, le.transform(le.classes_)))})")

# 3. One-Hot Encoding pour les variables catégorielles non binaires
print("\n3. One-Hot Encoding pour les variables multicatégories :")
colonnes_onehot = ['region', 'education_level', 'employment_status', 'loan_purpose']
colonnes_onehot = [col for col in colonnes_onehot if col in df_encoded.columns]

df_encoded = pd.get_dummies(df_encoded, columns=colonnes_onehot, prefix=colonnes_onehot, drop_first=True)
print(f"    {len(colonnes_onehot)} colonnes encodées en One-Hot")

print(f"\n Dataset après encodage : {df_encoded.shape[0]} lignes, {df_encoded.shape[1]} colonnes")


=== ÉTAPE 4-3 : Encodage des variables catégorielles ===

1. Encodage de la variable cible 'credit_risk' :
   - good → 1
   - bad → 0
    credit_risk encodée

2. Encodage des variables binaires :
    gender encodée (valeurs: {'F': np.int64(0), 'M': np.int64(1)})
    marital_status encodée (valeurs: {'divorced': np.int64(0), 'married': np.int64(1), 'single': np.int64(2), 'widowed': np.int64(3)})
    housing_status encodée (valeurs: {'family': np.int64(0), 'other': np.int64(1), 'own': np.int64(2), 'rent': np.int64(3)})
    consent_flag encodée (valeurs: {np.False_: np.int64(0), np.True_: np.int64(1)})

3. One-Hot Encoding pour les variables multicatégories :
    4 colonnes encodées en One-Hot

 Dataset après encodage : 3200 lignes, 52 colonnes


**5-PROTÉGER LES DONNÉES PERSONNELLES :**

**5-1-Supprimer les identifiants directs :**

In [6]:
print("\n=== ÉTAPE 5-1 : Suppression des identifiants directs ===\n")

# Liste des identifiants directs à supprimer
identifiants_directs = ['customer_id', 'full_name', 'email', 'phone', 'national_id']

# Vérifier quelles colonnes existent
colonnes_a_supprimer = [col for col in identifiants_directs if col in df_encoded.columns]
print(f"Colonnes à supprimer : {colonnes_a_supprimer}")

# Supprimer les colonnes
df_encoded = df_encoded.drop(columns=colonnes_a_supprimer)
print(f" {len(colonnes_a_supprimer)} colonnes supprimées")

print(f"Dataset après suppression : {df_encoded.shape[0]} lignes, {df_encoded.shape[1]} colonnes")


=== ÉTAPE 5-1 : Suppression des identifiants directs ===

Colonnes à supprimer : ['customer_id', 'full_name', 'email', 'phone', 'national_id']
 5 colonnes supprimées
Dataset après suppression : 3200 lignes, 47 colonnes


**5-2-Anonymiser les données financières**

In [7]:
print("\n=== ÉTAPE 5-2 : Anonymisation des données financières ===\n")

# Colonnes financières à anonymiser
colonnes_financieres = ['annual_income', 'savings_balance', 'checking_balance', 'loan_amount']

print("Anonymisation par normalisation (StandardScaler) :")
scaler = StandardScaler()

for col in colonnes_financieres:
    if col in df_encoded.columns:
        # Sauvegarder les valeurs originales pour référence (optionnel)
        df_encoded[col + '_original'] = df_encoded[col]

        # Normaliser les valeurs
        df_encoded[col + '_normalized'] = scaler.fit_transform(df_encoded[[col]])
        print(f"   {col} normalisée (moyenne=0, écart-type=1)")

        # Supprimer la colonne originale
        df_encoded = df_encoded.drop(columns=[col])

print(f"\n Dataset après anonymisation : {df_encoded.shape[0]} lignes, {df_encoded.shape[1]} colonnes")


=== ÉTAPE 5-2 : Anonymisation des données financières ===

Anonymisation par normalisation (StandardScaler) :
   annual_income normalisée (moyenne=0, écart-type=1)
   savings_balance normalisée (moyenne=0, écart-type=1)
   checking_balance normalisée (moyenne=0, écart-type=1)
   loan_amount normalisée (moyenne=0, écart-type=1)

 Dataset après anonymisation : 3200 lignes, 51 colonnes


**5-3-Pseudonymiser les quasi-identifiants :**

In [8]:
print("\n=== ÉTAPE 5-3 : Pseudonymisation des quasi-identifiants ===\n")
# 1. Agréger l'âge en tranches
if 'age' in df_encoded.columns:
    print("1. Agrégation de l'âge en tranches :")
    bins = [0, 25, 35, 45, 55, 100]
    labels = ['18-25', '26-35', '36-45', '46-55', '56+']
    df_encoded['age_group'] = pd.cut(df_encoded['age'], bins=bins, labels=labels, right=False)
    df_encoded = df_encoded.drop(columns=['age'])
    print(f"   Âge agréé en {len(labels)} tranches : {labels}")
    print(f"   Distribution :\n{df_encoded['age_group'].value_counts().sort_index()}")

# 2. Catégoriser les revenus (si non déjà normalisés)
print("\n2. Catégorisation des revenus :")
if 'annual_income_normalized' in df_encoded.columns:
    # Créer des catégories de revenus basées sur les percentiles
    df_encoded['income_category'] = pd.qcut(df_encoded['annual_income_normalized'],
                                             q=4,
                                             labels=['Très bas', 'Bas', 'Élevé', 'Très élevé'])
    print(f"   Revenus catégorisés en 4 classes")
    print(f"   Distribution :\n{df_encoded['income_category'].value_counts().sort_index()}")

print(f"\n Dataset après pseudonymisation : {df_encoded.shape[0]} lignes, {df_encoded.shape[1]} colonnes")


=== ÉTAPE 5-3 : Pseudonymisation des quasi-identifiants ===

1. Agrégation de l'âge en tranches :
   Âge agréé en 5 tranches : ['18-25', '26-35', '36-45', '46-55', '56+']
   Distribution :
age_group
18-25     332
26-35     829
36-45    1133
46-55     682
56+       224
Name: count, dtype: int64

2. Catégorisation des revenus :
   Revenus catégorisés en 4 classes
   Distribution :
income_category
Très bas      800
Bas           800
Élevé         800
Très élevé    800
Name: count, dtype: int64

 Dataset après pseudonymisation : 3200 lignes, 52 colonnes


**5-4-Masquage des données sensibles :**

In [9]:
print("\n=== ÉTAPE 5-4 : Masquage des données sensibles ===\n")

# Créer un identifiant générique pour les clients
if 'customer_id' in df.columns:
    print("1. Création d'un identifiant générique :")
    df_encoded['customer_reference'] = ['USER' + str(i).zfill(4) for i in range(1, len(df_encoded) + 1)]
    print(f"  Customer ID masqué : {df_encoded['customer_reference'].iloc[:5].tolist()} ...")

# Masquer les noms (si présents)
if 'full_name' in df.columns:
    print("\n 2. Masquage des noms :")
    df_encoded['full_name_masked'] = ['Client_' + str(i) for i in range(1, len(df_encoded) + 1)]
    print(f"  Noms masqués : {df_encoded['full_name_masked'].iloc[:5].tolist()} ...")

# Masquer les emails (si présents)
if 'email' in df.columns:
    print("\n 3. Masquage des emails :")
    df_encoded['email_masked'] = ['email_' + str(i) + '@masked.com' for i in range(1, len(df_encoded) + 1)]
    print(f"   Emails masqués : {df_encoded['email_masked'].iloc[:5].tolist()} ...")

print(f"\n Dataset après masquage : {df_encoded.shape[0]} lignes, {df_encoded.shape[1]} colonnes")


=== ÉTAPE 5-4 : Masquage des données sensibles ===

1. Création d'un identifiant générique :
  Customer ID masqué : ['USER0001', 'USER0002', 'USER0003', 'USER0004', 'USER0005'] ...

 2. Masquage des noms :
  Noms masqués : ['Client_1', 'Client_2', 'Client_3', 'Client_4', 'Client_5'] ...

 3. Masquage des emails :
   Emails masqués : ['email_1@masked.com', 'email_2@masked.com', 'email_3@masked.com', 'email_4@masked.com', 'email_5@masked.com'] ...

 Dataset après masquage : 3200 lignes, 55 colonnes


In [10]:
# ÉTAPE : CRÉER LE DATASET ML-READY

print("\n" + "="*60)
print("ÉTAPE 15 : CRÉATION DU DATASET ML-READY")
print("="*60)

print("\n 15.1 : Préparation du dataset final")

# Définir les colonnes utiles pour le modèle
colonnes_utiles = [
    # Quasi-identifiants pseudonymisés
    'age_group',
    'gender_encoded',
    'marital_status_encoded',
    'housing_status_encoded',
    'income_category',
    'customer_reference',

    # Données financières anonymisées
    'annual_income_normalized',
    'savings_balance_normalized',
    'checking_balance_normalized',
    'loan_amount_normalized',
    'loan_duration_months',
    'existing_credits',
    'past_due_count',
    'credit_history_score',
    'debt_ratio',

    # One-Hot encodées
    'loan_purpose_business',
    'loan_purpose_car',
    'loan_purpose_education',
    'loan_purpose_health',
    'loan_purpose_home_improvement',
    'loan_purpose_housing',
    'loan_purpose_other',

    # Variable cible
    'credit_risk_encoded'
]

# Ajouter les colonnes One-Hot existantes pour region, education, employment
for col in df_encoded.columns:
    if col.startswith(('region_', 'education_level_', 'employment_status_')):
        if col not in colonnes_utiles:
            colonnes_utiles.append(col)

# Garder uniquement les colonnes qui existent
colonnes_existantes = [col for col in colonnes_utiles if col in df_encoded.columns]
df_ml = df_encoded[colonnes_existantes]

print(f" Dataset ML-ready : {df_ml.shape[0]} lignes, {df_ml.shape[1]} colonnes")

# 15.2 : Vérification finale
print("\n 15.2 : Vérification finale")

print("  1. Statistiques descriptives :")
print(f"      - Moyenne d'âge : {df_ml['age_group'].value_counts().index[0] if 'age_group' in df_ml.columns else 'N/A'}")
print(f"      - Revenu moyen (normalisé) : {df_ml['annual_income_normalized'].mean():.2f}" if 'annual_income_normalized' in df_ml.columns else "")
print(f"      - Dette moyenne : {df_ml['debt_ratio'].mean():.2f}" if 'debt_ratio' in df_ml.columns else "")

print("\n  2. Distribution de la variable cible :")
if 'credit_risk_encoded' in df_ml.columns:
    print(f"      - 0 (good) : {(df_ml['credit_risk_encoded'] == 0).sum()} clients")
    print(f"      - 1 (bad) : {(df_ml['credit_risk_encoded'] == 1).sum()} clients")

print("\n  3. Valeurs manquantes :")
print(f"      - {df_ml.isnull().sum().sum()} valeurs manquantes totales")

print("\n  4. Aperçu du dataset :")
display(df_ml.head())

# 15.3 : Sauvegarder le dataset ML-ready
print("\n 15.3 : Sauvegarde du dataset ML-ready")

# Créer le dossier data/processed
os.makedirs('data/processed', exist_ok=True)

# Sauvegarder localement dans Colab
df_ml.to_csv('data/processed/credit_processed.csv', index=False)
print("   Dataset sauvegardé localement : data/processed/credit_processed.csv")


# TÉLÉCHARGER VERS L'ORDINATEUR LOCAL

print("\n" + "="*60)
print(" TÉLÉCHARGEMENT VERS VOTRE ORDINATEUR")
print("="*60)

print("\nChoisissez l'emplacement sur votre ordinateur pour sauvegarder le fichier :")

# Option 1 : Télécharger directement via le navigateur
print("\n OPTION 1 : Télécharger via le navigateur")
print("   Cliquez sur le bouton ci-dessous pour télécharger le fichier")
files.download('data/processed/credit_processed.csv')

print("\n Fichier téléchargé sur votre ordinateur !")

# Option 2 : Sauvegarder avec un nom personnalisé
print("\n" + "="*60)
print(" SAUVEGARDE AVEC NOM PERSONNALISÉ")
print("="*60)

print("\nVoulez-vous renommer le fichier avant de le sauvegarder ?")
print("Par défaut : credit_processed.csv")

# Proposer de renommer
custom_name = input("\nEntrez un nouveau nom (ou appuyez sur Entrée pour garder le nom par défaut) : ")

if custom_name.strip():
    custom_filename = custom_name.strip()
    if not custom_filename.endswith('.csv'):
        custom_filename += '.csv'
else:
    custom_filename = 'credit_processed.csv'

# Sauvegarder avec le nouveau nom
df_ml.to_csv(custom_filename, index=False)
print(f" Dataset sauvegardé sous : {custom_filename}")

# Télécharger avec le nouveau nom
print(f"\n Téléchargement de {custom_filename}...")
files.download(custom_filename)
print(f" {custom_filename} téléchargé sur votre ordinateur !")


ÉTAPE 15 : CRÉATION DU DATASET ML-READY

 15.1 : Préparation du dataset final
 Dataset ML-ready : 3200 lignes, 41 colonnes

 15.2 : Vérification finale
  1. Statistiques descriptives :
      - Moyenne d'âge : 36-45
      - Revenu moyen (normalisé) : 0.00
      - Dette moyenne : 0.65

  2. Distribution de la variable cible :
      - 0 (good) : 1744 clients
      - 1 (bad) : 1456 clients

  3. Valeurs manquantes :
      - 0 valeurs manquantes totales

  4. Aperçu du dataset :


,age_group,gender_encoded,marital_status_encoded,housing_status_encoded,income_category,customer_reference,annual_income_normalized,savings_balance_normalized,checking_balance_normalized,loan_amount_normalized,...,region_Trarza,education_level_master,education_level_phd,education_level_primary,education_level_secondary,employment_status_public_sector,employment_status_retired,employment_status_self_employed,employment_status_student,employment_status_unemployed
0,36-45,1,2,2,Élevé,USER0001,0.174551,1.040954,0.131306,0.630591,...,False,False,False,False,False,True,False,False,False,False
1,36-45,1,1,3,Élevé,USER0002,0.429863,0.081151,-1.279019,1.487241,...,False,False,False,False,True,False,False,True,False,False
2,26-35,0,1,3,Élevé,USER0003,0.418044,1.321549,-0.978143,-1.461699,...,False,False,False,False,False,False,False,True,False,False
3,46-55,0,0,3,Très bas,USER0004,-0.984885,-0.266905,0.048140,0.024276,...,False,False,False,True,False,False,False,True,False,False
4,46-55,1,1,2,Très bas,USER0005,-1.421001,-0.804513,-0.667361,-0.494273,...,False,False,False,False,True,False,False,False,True,False



 15.3 : Sauvegarde du dataset ML-ready
   Dataset sauvegardé localement : data/processed/credit_processed.csv

 TÉLÉCHARGEMENT VERS VOTRE ORDINATEUR

Choisissez l'emplacement sur votre ordinateur pour sauvegarder le fichier :

 OPTION 1 : Télécharger via le navigateur
   Cliquez sur le bouton ci-dessous pour télécharger le fichier


<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>


 Fichier téléchargé sur votre ordinateur !

 SAUVEGARDE AVEC NOM PERSONNALISÉ

Voulez-vous renommer le fichier avant de le sauvegarder ?
Par défaut : credit_processed.csv

Entrez un nouveau nom (ou appuyez sur Entrée pour garder le nom par défaut) : 
 Dataset sauvegardé sous : credit_processed.csv

 Téléchargement de credit_processed.csv...


<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

 credit_processed.csv téléchargé sur votre ordinateur !


In [11]:
"""
ÉTAPES : CALCUL DES HASH SHA-256 (Version simplifiée)
"""
import hashlib
import os
from google.colab import files
from datetime import datetime

print("="*60)
print("CALCUL DES HASH SHA-256")
print("="*60)

# Créer le dossier hashes
os.makedirs('hashes', exist_ok=True)

# Fonction pour calculer le hash
def calculate_sha256(file_path):
    sha256_hash = hashlib.sha256()
    with open(file_path, "rb") as f:
        for byte_block in iter(lambda: f.read(4096), b""):
            sha256_hash.update(byte_block)
    return sha256_hash.hexdigest()

# ÉTAPE 19 : Hash du dataset brut

print("\n ÉTAPE 19 : Hash du dataset brut")
print(" Upload du fichier brut (credit.csv ou data_credit_risk.csv) :")
uploaded = files.upload()
filename = list(uploaded.keys())[0]

# Calculer le hash
hash_brut = calculate_sha256(filename)

# Sauvegarder
with open('hashes/raw_dataset_hash.txt', 'w') as f:
    f.write("HASH SHA-256 - DATASET BRUT\n")
    f.write("="*50 + "\n")
    f.write(f"Fichier : {filename}\n")
    f.write(f"Date : {datetime.now().strftime('%Y-%m-%d %H:%M:%S')}\n")
    f.write(f"Hash : {hash_brut}\n")

print(f"\n Hash brut : {hash_brut}")
print(f" Sauvegardé : hashes/raw_dataset_hash.txt")
files.download('hashes/raw_dataset_hash.txt')

# ÉTAPE 20 : Hash du dataset traité

print("\n" + "="*60)
print("ÉTAPE 20 : Hash du dataset traité")

processed_file = 'data/processed/credit_processed.csv'

if os.path.exists(processed_file):
    hash_traite = calculate_sha256(processed_file)

    with open('hashes/processed_dataset_hash.txt', 'w') as f:
        f.write("HASH SHA-256 - DATASET TRAITÉ\n")
        f.write("="*50 + "\n")
        f.write(f"Fichier : {processed_file}\n")
        f.write(f"Date : {datetime.now().strftime('%Y-%m-%d %H:%M:%S')}\n")
        f.write(f"Hash : {hash_traite}\n")

    print(f"\n Hash traité : {hash_traite}")
    print(f"Sauvegardé : hashes/processed_dataset_hash.txt")
    files.download('hashes/processed_dataset_hash.txt')
else:
    print("\n Dataset traité non trouvé !")
    print("Upload du fichier traité (credit_processed.csv) :")
    uploaded_processed = files.upload()
    processed_filename = list(uploaded_processed.keys())[0]

    hash_traite = calculate_sha256(processed_filename)

    with open('hashes/processed_dataset_hash.txt', 'w') as f:
        f.write("HASH SHA-256 - DATASET TRAITÉ\n")
        f.write("="*50 + "\n")
        f.write(f"Fichier : {processed_filename}\n")
        f.write(f"Date : {datetime.now().strftime('%Y-%m-%d %H:%M:%S')}\n")
        f.write(f"Hash : {hash_traite}\n")

    print(f"\n Hash traité : {hash_traite}")
    print(f"Sauvegardé : hashes/processed_dataset_hash.txt")
    files.download('hashes/processed_dataset_hash.txt')

print("\n" + "="*60)
print("TOUS LES HASH ONT ÉTÉ TÉLÉCHARGÉS !")
print("="*60)

CALCUL DES HASH SHA-256

 ÉTAPE 19 : Hash du dataset brut
 Upload du fichier brut (credit.csv ou data_credit_risk.csv) :


Saving data_credit_risk.csv to data_credit_risk (5).csv

 Hash brut : 6632f41cd72dca5e15896b66ebe53ba603f58f8e9cb85accef7ef20b6df47785
 Sauvegardé : hashes/raw_dataset_hash.txt


<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>


ÉTAPE 20 : Hash du dataset traité

 Hash traité : b3566267e18437effccb44d4cf818908a9bee4a1c3e773303090d069fbf39d32
Sauvegardé : hashes/processed_dataset_hash.txt


<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>


TOUS LES HASH ONT ÉTÉ TÉLÉCHARGÉS !
